### ISSUE TITLE
[drag rows] Row drag and drop

### ISSUE THREAD

**Issue by carloshpds** [ag-grid/ag-grid#802]
Is there any update for this feature?

**Comment by ceolter** [comment]
nope. work for this is not scheduled.

**Comment by carloshpds** [comment]
Thanks for the quick answer @ceolter ;)

**Comment by ceolter** [comment]
np. sorry for not the answer you were looking for. if it does help, i did expose the row via the callback processRowPostCreate(), i know others are using this to implement their own DnD.

**Comment by carloshpds** [comment]
@ceolter, just to keep you up to date.
`processRowPostCreate()` works just fine using AngularJS + [ngDraggable](https://github.com/fatlinesofcode/ngDraggable) to implement row DnD. :)
Where would be the better place to post an example?

**Comment by DaVarga** [comment]
@carloshpds could you please drop an working example in [plunker](https://plnkr.co/)? That would be awesome.
Thanks.

**Comment by ceolter** [comment]
i can include it in my docs? if you post on plunker or similar i can include a link from the docs pages.

**Comment by carloshpds** [comment]
I will make a plunk this weekend and post the link here :smile:

**Comment by Brocco** [comment]
@carloshpds I would love to see your demo/plunker when you're done with it, my client is requesting this feature and I'm leaning towards using ag-grid

**Comment by DaVarga** [comment]
+1

**Comment by carloshpds** [comment]
Oooh guys, sorry about that. I was too busy these days. But i've started to write a pen at codepen. It's really simple. I'll post it here as soon as possible :)

**Comment by carloshpds** [comment]
@DaVarga @Brocco @ceolter 
Hey guys, what's up?
Finally, I had time to write this example!!! \o/
So here is the [link](https://embed.plnkr.co/qLy0EX/) and would be nice to have some feedback about it :)
Really really sorry for taking so long!

**Comment by DaniVarga** [comment]
Hey, thanks a lot. Ill have a closer look on Monday. Have a nice weekend.

**Comment by fabiogallo-blueprint** [comment]
@carloshpds Thanks a lot for your example!
I used it as a base for adding d&d support for ag-grid as treeview component.
It stil has a couple of bugs but here is the version I came up with: https://plnkr.co/HfPXsl

**Comment by jukka-varis** [comment]
In version 5.x adding directive onto row is not working, because `processRowPostCreate` function is executed after compiling for angular. In 4.x version compiling is done as last thing in `RenderedRow` init phase.

I tested it by adding this
`if (this.scope) {
	this.$compile(this.eBodyRow)(this.scope);
}`
to the end of `init` function of `RenderedRow` and drag & drop suggested by @carloshpds started working in 5.4.0 version.

**Comment by shivekundefined** [comment]
@jukka-varis , Please give us some plnker or any example that how you have done this.

**Comment by le4dface** [comment]
Has anyone succeed to re-implementing the drag and drop solution for 4.x in 6.1.x? I added $compile($row)($scope) to the end of my processRowPostCreate() callback and this partially work. The list does not reorder on a drop event, however.

**Comment by fdmota** [comment]
I am also looking for a solution for this in version 6.x.

**Comment by Miteshdv** [comment]
Can  processRowPostCreate   be used to drag and drop in react ag-grid? please let me know thanks

**Comment by ceolter** [comment]
tracked by  AG-116
https://www.ag-grid.com/ag-grid-pipeline/

**Comment by flakyfilibuster** [comment]
Thanks for adding this to the release pipeline @ceolter!

I've seen, that it is added to the "Scheduled for Dev - should be in a release soon" table. 
This question might err on the side of annoying, however is there a ballpark estimate on when it is going to be released?

Thanks in advance!

**Comment by kvtm-annukampa** [comment]
What if i have to drag and drop rows in ag-grid version 3.x .

**Comment by shivekundefined** [comment]
@ceolter  When this will be completed? I am waiting for this feature to be implemented for decades.

**Comment by PTC-JoshuaMatthews** [comment]
Just adding that I am hoping for this feature as well!

**Comment by lobo-tuerto** [comment]
What's the current recommended way that I should try if I want to add drag and drop capabilities to ag-grid?

**Comment by AnuragSwarnkar** [comment]
I am using angularCompileRows=true as options and initialize the ag-grid using new agGrid(<divid>);
but getting error of $new of null.

**Comment by jathri** [comment]
Looking forward to see this feature too. Will it be available for free users?

**Comment by vgramagin** [comment]
Hi,

the example on plunker works perfectly, but once I change ag-grid version to 13.1.2 it stops. Is there any workaround available for the latest version?

@carloshpds do you have an updated example by any chance? That would be so great

**Comment by marc314** [comment]
I came up with a solution if you are using React JS. You would need to use a little jQuery ui for the draggable component. Here's a working sample inspired from the angular solution.
```
// style.css
.ag-row {
  cursor: move;
}

.ag-row.drag-enter:not(.drag-over) {
  border: 2px solid blue;
  z-index: 1;
}

.ag-row.drag-over {
  opacity: 0.5;
  cursor: move !important;
}


```
import React, { Component } from "react";
import { AgGridReact } from "ag-grid-react";
import $ from "jquery";
import "jquery-ui";
import "./style.css";

function immutableSplice(arr, start, deleteCount, ...items) {
  return [...arr.slice(0, start), ...items, ...arr.slice(start + deleteCount)];
}

export default class extends Component {
  constructor(props) {
    super(props);

    this.state = {
      columnDefs: this.createColumnDefs(),
      rowData: this.createRowData()
    };
  }

  onGridReady = params => {
    this.gridApi = params.api;
    this.columnApi = params.columnApi;

    this.gridApi.sizeColumnsToFit();
  };

  createColumnDefs() {
    return [{ headerName: "Make", field: "make" }];
  }

  createRowData() {
    return [
      { make: "A", model: "Celica", price: 35000 },
      { make: "B", model: "Mondeo", price: 32000 },
      { make: "C", model: "Boxter", price: 72000 },
      { make: "D", model: "m3", price: 80000 }
    ];
  }
  processRowPostCreate = params => {
    let self = this;
    if (params.eRow) {
      let $row = params.eRow;
      $row.draggable = true;
      $($row).on("dragstart", function(event) {
        $(this).addClass("drag-over");
        const node = params.node;
        self.dragFromIndex = node.rowIndex;
      });
      $($row).on("dragover", function(event) {
        event.preventDefault();
        event.stopPropagation();
        $(this).addClass("drag-enter");
      });

      $($row).on("dragleave", function(event) {
        event.preventDefault();
        event.stopPropagation();
        $(this).removeClass("drag-enter");
      });

      $($row).on("dragend", function(event) {
        event.preventDefault();
        event.stopPropagation();
        delete self.dragFromIndex;
        $(".ag-row").removeClass("drag-over");
        $(".ag-row").removeClass("drag-enter");
      });
      $($row).on("drop", function(event) {
        event.preventDefault();
        event.stopPropagation();
        const node = params.node;
        self.onDropRow(self.dragFromIndex, node.rowIndex);
      });
    }
  };

  onDropRow = (srcIdx, destIdx) => {
    if (srcIdx === destIdx) return;

    var rows = immutableSplice(this.state.rowData, srcIdx, 1);
    rows = immutableSplice(rows, destIdx, 0, this.state.rowData[srcIdx]);
    this.setState(
      {
        rowData: rows
      },
      () => this.gridApi.setRowData(rows)
    );
  };

  render() {
    let containerStyle = {
      height: 200,
      width: 500
    };

    return (
      <div style={containerStyle} className="ag-fresh">
        <h1>Simple ag-Grid React Example</h1>
        <AgGridReact
          // properties
          columnDefs={this.state.columnDefs}
          rowData={this.state.rowData}
          processRowPostCreate={this.processRowPostCreate}
          // events
          onGridReady={this.onGridReady}
        />
      </div>
    );
  }
}

```

Hopefully this helps someone.

Best,
Marc

**Comment by philly-vanilly** [comment]
@ceolter Glad to see row drag and drop was not forgotten in the road map and you guys reviewed it recently. Will this a free or enterprise feature, have you decided yet?

**Comment by matte00** [comment]
I need your help! I'm trying implement drag and drop from ag-grid (rows) to angular-tree-component. Angular-tree-component provide a method for drag an external element into the tree. But for this, you must add a directive inside the template. 
Example:
`<p [treeDrag]="{name: 'first'}" [treeDragEnabled]="true">Drag me!</p>`

I can't do it because i need to drag an ag-grid's row where i cannot add custom template directive (or i can do it??). Can you help me? Any ideas?

**Comment by suparnadk** [comment]
Hi, AG-116 is removed from the pipeline I see , any idea which release would it be included in?

**Comment by makinggoodsoftware** [comment]
Hi, 

Thank you for raising this issue for us, at this moment we are reviewing how we process Feature Requests/questions that come through from GitHub.

As I am sure you are aware, we offer two products:

ag-Grid - this is entirely free to use in line with our MIT license
ag-Grid Enterprise - our commercial version which funds the development of both versions.

We currently have a very full pipeline of work which can be viewed here: https://www.ag-grid.com/ag-grid-pipeline/

This will be our priority for the coming months and as a result, we don’t have capacity to take feature requests/questions from GitHub, but you might find that yours has already been raised in our pipeline.

We will do a periodic review of GitHub issues for any bugs raised and these can also be tracked on our pipeline page.

Thanks for your understanding.
ag-Grid Team.

**Comment by JohnMasterson** [comment]
AG-116 was included in Version 16.

Regards, 

John

**Comment by panda97331** [comment]
https://www.ag-grid.com/javascript-grid-row-dragging/

**Comment by pabacham** [comment]
Is there a way to drag a row within a group (say reorder)? Because in this example we can drag only between different groups https://www.ag-grid.com/javascript-grid-row-dragging/#example-dragging-with-row-groups

**Comment by EugenMaevskiy** [comment]
Have same question.
Maybe someone could advice?

**Comment by mxswat** [comment]
Have same question. As pabacham and EugenMaevskiy.
Any update?

**Comment by raghavyadavm** [comment]
Same Question as pabacham. Any update?

**Comment by cherry-aggrid** [comment]
var rowData = ["abc","def","ghi"]; // some random row data.
processRowPostCreate: function onRowPostCreate(params) {
	var $row = angular.element(params.eRow);
	var dropSuccessCallbackAsString = "onDropRow(" + params.rowIndex + ", $data, $event, $target)";
	$row.attr('ng-drop', true);
	$row.attr('ng-drop-success', dropSuccessCallbackAsString);
	$row.attr('ng-drag', true);
	$row.attr('ng-drag-data', 'data');
}. 

function onDropRow(newIndex, data, event, target) {
var oldIndex = rowData.indexOf(data);
gridOptions.splice(oldIndex, 1);
gridOptions.splice(newIndex, 0, data);
gridOptions.api.setRowData(rowData);
}

this is the code I used for dragging and dropping the row in my ag grid using angularJS. But the issue is, it doesn't work.
In params.eRow  eRow is undefined. I tried a lot, but it is still undefined. Any help is appreciable.

Thanks
